packages/discovery-react-components/src/components/DocumentPreview/components/PdfViewer/PdfViewerTextLayer.tsx

```diff
@@ -0,0 +1,207 @@
+import React, { FC, useEffect, useRef, useState } from 'react';
+import cx from 'classnames';
+import PDFJSLib, { PDFPageProxy, PDFPageViewport, TextContent, TextContentItem } from 'pdfjs-dist';
+import { EventBus } from 'pdfjs-dist/lib/web/ui_utils';
+import { TextLayerBuilder } from 'pdfjs-dist/lib/web/text_layer_builder';
+
+const { RenderingCancelledException } = PDFJSLib as any;
+
+interface Props {
+  className?: string;
+
+  /**
+   * PDF page from pdfjs
+   */
+  loadedPage: PDFPageProxy | null | undefined;
+
+  /**
+   * Page number, starting at 1
+   */
+  page: number;
+
+  /**
+   * Zoom factor, where `1` is equal to 100%
+   */
+  scale: number;
+
+  /**
+   * Callback for text layer info
+   */
+  setTextLayerInfo?: (info: PdfTextLayerInfo | null) => any;
+}
+
+export type PdfTextLayerInfo = {
+  /**
+   * PDF text content
+   */
+  textContent: TextContent & {
+    styles: { [styleName: string]: CSSStyleDeclaration };
+  };
+
+  /**
+   * Text span DOM elements rendered on the text layer
+   */
+  textDivs: HTMLElement[];
+
+  /**
+   * Pdf page viewport used to render text items
+   */
+  viewport: PDFPageViewport;
+
+  /**
+   * Page number, starting at 1
+   */
+  page: number;
+};
+
+const PdfViewerTextLayer: FC<Props> = ({
+  className,
+  loadedPage,
+  page = 1,
+  scale = 1,
+  setTextLayerInfo: setTextLayerInfoCallback = () => {}
+}) => {
+  const textLayerRef = useRef<HTMLDivElement>(null);
+  const textLayerDiv = textLayerRef.current;
+
+  const [textRenderInfo, setTextRenderInfo] = useState<{
+    page: number;
+    scale: number;
```

Comment:
These are incoming props. There's no reason to store them in state.

Reply:
I had to do this tweak to make sure all the variables are updated in synchronous. As we have to use async method to load PDF text content, there will be timing gap between the time `loadedPage` is updated and the time `textContent` is updated. It can triggers text rendering in the middle of updating states. 

Maybe we can overcome it by having one large async method, but I thought that it's worth splitting`getTextConten` from PDF and `renderTextContent`. I appreciate your suggestion.